hammerblade/torch/kernel/bsg_tensor.hpp

```diff
@@ -13,4 +31,100 @@ typedef struct {
 #endif
 } bsg_tensor_t;
 
+typedef struct {
+  uint32_t N;
+#ifdef HB_EMUL
+  uint64_t data;
+#else
+  uint32_t data;
+#endif
+} bsg_vector_t;
+
+// =========================================================
+// Device Tensor classes
+//
+// Wrapper classes around device tensor structs to provide
+// convenience operations. This runs on a tiny RISC-V processor 
+// on the device, so be careful about using dynamic memory 
+// allocation.
+// =========================================================
+
+class BSGTensor {
+  private:
+    uint32_t N;
+    uint32_t dims;
+    uint32_t* strides;
+    float* data;
+
+  public:
+    BSGTensor(bsg_tensor_t* t) :
+      N(t->N),
+      dims(t->dims),
+      strides((uint32_t*) ((intptr_t) t->strides)),
+      data((float*) ((intptr_t) t->data)) {}
+
+    int size() {
+      return N;
+    }
+
+    uint32_t dim(uint32_t d) {
+      if(d >= dims) {
+        bsg_printf("BSGTensor error: dimesnion must be less than %d\n",
+            dims);
+      }
+
+      uint32_t dim;
+
+      if(d == 0) {
+        dim =  N / strides[0];
```

Comment:
> Actually, I think I will add a new data structure for offloading a TensorIterator.

The reason is that for tensors constructed out of TensorIterators, `strides[i]` can be 0. If we don't careful enough we will have a floating point exception here. So I think I'll add a new set of structs/classes just for iterators.

Reply:
Sounds good!

Couple of things I understood about TensorIterators:
- Those are exclusively for element-wise operations and hence are always 1-d tensors.
- Strides means number of bytes probably to distinguish float vs bfloat16 on GPUs.